// @ngx-loading-bar/http-client/@ngx-loading-bar/http-client.es5.js
import { LoadingBarModule, LoadingBarService } from '@ngx-loading-bar/core';
import { Injectable, NgModule } from '@angular/core';
import { HTTP_INTERCEPTORS, HttpClientModule, HttpEventType } from '@angular/common/http';
import { _do as _do$1 } from 'rxjs/operator/do';
import { _finally as _finally$1 } from 'rxjs/operator/finally';
var LoadingBarInterceptor = /** @class */ (function () {
    /**
     * @param {?} loadingBar
     */
    function LoadingBarInterceptor(loadingBar) {
        this.loadingBar = loadingBar;
    }
    /**
     * @param {?} req
     * @param {?} next
     * @return {?}
     */
    LoadingBarInterceptor.prototype.intercept = function (req, next) {
        var _this = this;
        // https://github.com/angular/angular/issues/18155
        var /** @type {?} */ meta = ((req)).metadata || {};
        if (meta && meta.ignoreLoadingBar === true) {
            return next.handle(req);
        }
        var /** @type {?} */ started = false;
        return _finally$1.call(_do$1.call(next.handle(req), function (event) {
            if (!started && event.type === HttpEventType.Sent) {
                started = true;
                _this.loadingBar.start();
            }
            else if (started && event.type === HttpEventType.Response) {
                started = false;
                _this.loadingBar.complete();
            }
        }), function () { return started && _this.loadingBar.complete(); });
    };
    return LoadingBarInterceptor;
}());
LoadingBarInterceptor.decorators = [
    { type: Injectable },
];
/**
 * @nocollapse
 */
LoadingBarInterceptor.ctorParameters = function () { return [
    { type: LoadingBarService, },
]; };
var LoadingBarHttpClientModule = /** @class */ (function () {
    function LoadingBarHttpClientModule() {
    }
    return LoadingBarHttpClientModule;
}());
LoadingBarHttpClientModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    HttpClientModule,
                    LoadingBarModule.forRoot(),
                ],
                exports: [
                    HttpClientModule,
                    LoadingBarModule,
                ],
                providers: [
                    { provide: HTTP_INTERCEPTORS, useClass: LoadingBarInterceptor, multi: true },
                ],
            },] },
];
/**
 * @nocollapse
 */
LoadingBarHttpClientModule.ctorParameters = function () { return []; };
/**
 * Generated bundle index. Do not edit.
 */
export { LoadingBarHttpClientModule, LoadingBarInterceptor as ɵa };
