// @ngx-loading-bar/http-client/@ngx-loading-bar/http-client.js
import { LoadingBarModule, LoadingBarService } from '@ngx-loading-bar/core';
import { Injectable, NgModule } from '@angular/core';
import { HTTP_INTERCEPTORS, HttpClientModule, HttpEventType } from '@angular/common/http';
import { _do as _do$1 } from 'rxjs/operator/do';
import { _finally as _finally$1 } from 'rxjs/operator/finally';

class LoadingBarInterceptor {
    /**
     * @param {?} loadingBar
     */
    constructor(loadingBar) {
        this.loadingBar = loadingBar;
    }
    /**
     * @param {?} req
     * @param {?} next
     * @return {?}
     */
    intercept(req, next) {
        // https://github.com/angular/angular/issues/18155
        const /** @type {?} */ meta = ((req)).metadata || {};
        if (meta && meta.ignoreLoadingBar === true) {
            return next.handle(req);
        }
        let /** @type {?} */ started = false;
        return _finally$1.call(_do$1.call(next.handle(req), (event) => {
            if (!started && event.type === HttpEventType.Sent) {
                started = true;
                this.loadingBar.start();
            }
            else if (started && event.type === HttpEventType.Response) {
                started = false;
                this.loadingBar.complete();
            }
        }), () => started && this.loadingBar.complete());
    }
}
LoadingBarInterceptor.decorators = [
    { type: Injectable },
];
/**
 * @nocollapse
 */
LoadingBarInterceptor.ctorParameters = () => [
    { type: LoadingBarService, },
];

class LoadingBarHttpClientModule {
}
LoadingBarHttpClientModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    HttpClientModule,
                    LoadingBarModule.forRoot(),
                ],
                exports: [
                    HttpClientModule,
                    LoadingBarModule,
                ],
                providers: [
                    { provide: HTTP_INTERCEPTORS, useClass: LoadingBarInterceptor, multi: true },
                ],
            },] },
];
/**
 * @nocollapse
 */
LoadingBarHttpClientModule.ctorParameters = () => [];

/**
 * Generated bundle index. Do not edit.
 */

export { LoadingBarHttpClientModule, LoadingBarInterceptor as ɵa };
